feat(eslint-plugin): add rule [strict-void-return]

phaux opened the pull request:
## PR Checklist

- [x] Addresses an existing open issue:
  - fixes #2988 
  - fixes #1744
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

So basically I implemented the void checking as requested by #2988. For every checked node (arguments, assignments, returns, etc) I take the actual function type and the contextual function type and compare return types. Only object method shorthand required slightly different logic. 

That already worked pretty well, but I also found #1744 and decided to include it in this rule as well, since I already had a similar thing implemented for object shorthand methods. I added this as an option `considerBaseClass` and `considerImplementedInterfaces`, enabled by default.

Then I noticed that callback for `addEventListener` is not detected as void context. That's because it has another signature where the callback can return `any`. I was stuck on this for a long time. Ultimately I looked at how `no-misused-promises` does this and implemented something similar as `considerOtherSignatures` option, enabled by default.

At this point this rule already did everything `no-misused-promises`'s `checkVoidReturn` did, but better. It doesn't have problems like #8054 or #8739. Maybe it's worth splitting `no-misused-promises` into 3 separate rules in the future?

I also added many autofixes and suggestions. They are possible when the provided function is a function literal and we can inspect its body. Some of them are the same as in `no-confusing-void-expression` so I moved them into utils. It might make sense to change some autofixes into suggestions instead so they don't accidentally remove a big chunk of code. Let me know if that's a good idea.

The biggest feature is automatic suggestions which I and probably others had to type manually a thousand of times:

```js
takesCallback(async () => {
  block;
});
``` 

into 

```js
takesCallback(() => {
  (async () => {
    block;
  })().catch(err => {});
});
```

or

```js
// eslint-disable-next-line @typescript-eslint/no-misused-promises
takesCallback(async () => {
  try {
    block;
  } catch {}
});
```

To allow the second suggestion without the need of ignoring the line I added the option `allowReturnPromiseIfTryCatch`. It's just a simple extra `if` near the end of the long routine that checks everything that could be wrong in the function body. I hope it can stay.

JoshuaKGoldberg commented:
👋 Exciting PR, really looking forward to the rule! Just marking as a draft because there are unit test failures. This keeps getting me excited when it pops up in my notifications 😄. Let us know if you want to talk or ask questions about any of them.

phaux commented:
Fixed and I'm not planning any more changes so I'm undrafting it I guess.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,991 @@
+            );
+          const argExpectedReturnTypes = funcSignatures
+            .map(s => s.parameters[argIdx])
+            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- Indexing can return undefined
+            .filter(param => param != null)
```
[Style] Minor nitpick, but helps get rid of the comment 😄 

```suggestion
            .filter(Boolean)
```

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,991 @@
+import assert from 'node:assert';
```
[Refactor] Rules can run in environments other than Node. But, even if we did want to take a dependency on `node:assert`, the kinds of assertions in this rule aren't what we'd do. Lint rules run in a controlled enough environment (known AST inputs) that we should be able to handle all cases without crashing. 

If something is known to be a specific type and there's no good way to tell the type system that, use an `as`.

Aside: it's common that the type system is actually correct when it's yelling about a node being the wrong type. Or at least - it's correct per the AST's types. If it's yelling and you think it shouldn't then that might be a bug in the AST types? For example, in the `if (funcNode.generator)` check, `funcNode.body` is typed as `TSESTree.BlockStatement | TSESTree.Expression`... even though if `funcNode.generator` is true, we know it's not an arrow expression. A bug in the AST types maybe?

phaux replied:
Why not add `assert` to utils like `nullThrows`?

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,991 @@
+      },
+      ObjectExpression: (node): void => {
+        for (const propNode of node.properties) {
+          if (propNode.type !== AST_NODE_TYPES.SpreadElement) {
```
[Testing] If I remove this `if` and just leave its body, all unit tests still pass. That means there's at least a missing test case.

phaux replied:
That's because technically this condition is not necessary, because if it's a spread argument then the contextual type wont be a void function. I don't know how to write a test for this. Is it even possible?

JoshuaKGoldberg replied:
Oop, sorry I missed this question. Looking at the non-spread test case that hits this line:

```ts
declare function cb(): void;
const foo: Record<string, () => void> = {
  cb1: cb,
  cb2: cb,
};
```

We'd be able to trigger it by switching one of those properties to a spread:

```ts
declare function cb(): void;
const foo: Record<string, () => void> = {
  cb1: cb,
  ...{},
};
```

phaux replied:
Done

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,991 @@
+        }
+      },
+      ArrowFunctionExpression: (node): void => {
+        if (node.body.type !== AST_NODE_TYPES.BlockStatement) {
```
[Testing] If I remove this `if` and just leave its body, all unit tests still pass. That means there's at least a missing test case.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,991 @@
+    const checker = parserServices.program.getTypeChecker();
+
+    return {
+      'CallExpression, NewExpression': (
+        node: TSESTree.CallExpression | TSESTree.NewExpression,
+      ): void => {
+        checkFunctionCallNode(node);
+      },
```
[Style] Nit: these can be inlined. Anything that just curries a single param ✨ 

```suggestion
      'CallExpression, NewExpression': checkFunctionCallNode,
```

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,991 @@
+        for (const elemNode of node.elements) {
+          if (
+            elemNode != null &&
+            elemNode.type !== AST_NODE_TYPES.SpreadElement
```
[Testing] If I remove this `&& elemNode.type !== AST_NODE_TYPES.SpreadElement`, all unit tests still pass. That means there's at least a missing test case.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,991 @@
+        }
+      },
+      AssignmentExpression: (node): void => {
+        if (['=', '||=', '&&=', '??='].includes(node.operator)) {
```
[Testing] If I remove this `if` and just leave its body, all unit tests still pass. That means there's at least a missing test case.

JoshuaKGoldberg replied:
This is still the case.

phaux replied:
Added a test but it doesn't matter:

```ts
declare let foo: () => void;
foo += () => 1;
```

still doesn't fail with condition removed.

JoshuaKGoldberg commented on packages/eslint-plugin/src/util/getBaseTypesOfClassMember.ts:
```diff
@@ -0,0 +1,39 @@
+export function* getBaseTypesOfClassMember(
+  checker: ts.TypeChecker,
+  memberTsNode: ts.PropertyDeclaration | ts.MethodDeclaration,
+  heritageToken: ts.SyntaxKind.ExtendsKeyword | ts.SyntaxKind.ImplementsKeyword,
```
[Refactor] Both places that call `getBaseTypesOfClassMember` look roughly like:

```ts
for (const { baseMemberType } of util.getBaseTypesOfClassMember(checker, propTsNode, ts.SyntaxKind.ExtendsKeyword) {
    if (isVoidReturningFunctionType(baseMemberType)) {
      reportIfNonVoidFunction(propNode.value);
  }
}

for (const { baseMemberType } of util.getBaseTypesOfClassMember(checker, propTsNode, ts.SyntaxKind.ExtendsKeyword) {
    if (isVoidReturningFunctionType(baseMemberType)) {
      reportIfNonVoidFunction(propNode.value);
  }
}
```

There's some duplication there. I think the `heritageToken` parameter can be removed, to have the function always check both in one call, rather than one each in two calls.

phaux replied:
Done

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,991 @@
+              memberName,
+              className,
+              baseName,
+            });
+          }
```
[Bug] It's kind of unusual to report twice in one place. Plus, for better performance, it's nice to not have to do checks later on once there's already a known report. For here and elsewhere, let's skip later reports as soon as one is made:

```suggestion
            });
            return;
          }
```

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/strict-void-return.mdx:
> At this point this rule already did everything `no-misused-promises`'s `checkVoidReturn` did, but better. It doesn't have problems #8054 or #8739. Maybe it's worth splitting `no-misused-promises` into 3 separate rules in the future? (this one being one of them)

This is a great question. `no-misused-promises` was already largely overlapped by `no-unnecessary-condition`. This new `strict-void-return` pretty much takes on the rest of `no-misused-promises`, making `no-misused-promises` redundant if you have both `no-unnecessary-condition` and `strict-void-return`... 

I'd be in favor of deprecated `no-misused-promises` in favor of using `no-unnecessary-condition` + `strict-void-return`. The only benefit I can think of for `no-misused-promises` would be projects that want to _only_ apply the checks for Promises... Maybe these two rules could each be given some kind of _"only check Promises"_ option?

Also of note is that `no-misused-promises`'s `checkVoidReturns` is pretty configurable. Maybe, if this rule is to replace `no-misused-promises`, it'd be useful to have each of those configurable options? Or, on the other hand, maybe those options are holdovers that real-world don't generally use? Investigation needed. I think those options can be a followup & shouldn't block this PR.

What do you think?

Also cc: @typescript-eslint/triage-team in general, and @kirkwaiblinger + @alythobani from #8765.

kirkwaiblinger replied:
Maybe a lame response, but is there a compelling reason not to land this first, then consider the no-misused-promises deprecation, and which options we might need to port or create in order to do so, afterwards?

Just thinking, deprecating no-misused-promises might have some strings attached, such as some nontrivial updating of the docs in no-floating-promises that explain how to lint against promise antipatterns outside of ExpressionStatements.

JoshuaKGoldberg reviewed (changes requested):
OK! Very happy to have finally made it to this PR: it's a great piece of work. As you said, it covers a heck of a lot of functionality, and does so in ways that are really solid compared to previous approaches. Fantastic! 👏 

It's also a _lot_ of code that was hard to read through. For an initial version of the rule, I think nuance around suggestions aren't necessary. And the fixers would need to be suggestions given they change code behavior.

I left requests for simplification through the code: for messages, options, and the suggestions.

But, my advice would be to hold off applying that large set of removals until the conversation in _(will switch to link)_ is resolved. The consensus might end up being that the options are good and useful after all.

![Keanu Reaves as Neo in The Matrix saying 'whoa'](https://github.com/user-attachments/assets/ee7fd804-9ad1-4885-8bdf-34bfba0b22c6)

phaux commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,991 @@
+        // Report a generic error.
+        return context.report({
+          node: funcNode,
+          messageId: `nonVoidFuncIn${msgId}`,
+          data,
```
I considered having context as an argument to messages like `"Unexpected value returned in {context}..."`, but then the context would be untranslatable, which I think is the whole point of using message IDs, no? 

Should I change it to taking context like `` `argument for ${funcName}` `` ?

JoshuaKGoldberg replied:
That seems reasonable to me!

alythobani reviewed (commented):
Very cool PR, great job sticking it out! Left some thoughts/questions, hope they're helpful

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,991 @@
+          checkExpressionNode(node.argument, 'Return');
+        }
+      },
+      PropertyDefinition: (node): void => {
+        checkClassPropertyNode(node);
+      },
+      MethodDefinition: (node): void => {
+        checkClassMethodNode(node);
+      },
```
I'd say let's wait on that, yeah. Maybe in a future PR.

alythobani replied:
Including for `no-misused-promises`?

alythobani replied:
Separate from this PR for sure, but I meant do we want to postpone that indefinitely until we have a better idea of what we're going to do with `no-misused-promises`?

JoshuaKGoldberg replied:
Oh, gotcha - no I think that's a different thing. Whether a rule properly checks index signatures is a _bug_ scoped _with the rule_. High-level direction discussions like this might change what rule(s) exist or where they look, but don't really impact small bugs like that.

phaux commented on packages/eslint-plugin/docs/rules/strict-void-return.mdx:
````diff
@@ -0,0 +1,436 @@
+
+```ts
+const bad: () => void = () => 2137;
+const func = Math.random() > 0.1 ? bad : prompt;
````
I tried

phaux commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,991 @@
+      considerImplementedInterfaces: true,
+      allowReturnPromiseIfTryCatch: true,
+      allowReturnUndefined: true,
+      allowReturnNull: true,
```
Makes sense

kirkwaiblinger commented on packages/eslint-plugin/docs/rules/strict-void-return.mdx:
````diff
@@ -0,0 +1,496 @@
+```ts
+const getNothing: () => void = () => 2137;
+const getString: () => string = () => 'Hello';
+const getStringMaybe = Math.random() > 0.1 ? getNothing : getString;
+const maybeString = getStringMaybe();
````
(praise) This version of the example is much easier to follow. Thanks! 🙂 

(nit) why not inline the `getStringMaybe` variable still?
```suggestion
const maybeString = Math.random() > 0.1 ? getNothing() : getString();
```

kirkwaiblinger commented on packages/eslint-plugin/docs/rules/strict-void-return.mdx:
```diff
@@ -0,0 +1,496 @@
+
+### Return type unsafety
+
+Passing a value-returning function in a place expecting a void function can be unsound.
```
What do you think of a little more verbosity here? 

If I were writing it, I might do something like this...

````suggestion
Passing a value-returning function in a place expecting a void function can be unsound. TypeScript generally treats the `void` type as though it has the same runtime behavior as `undefined`, so this pattern will cause a mismatch between the runtime behavior and the types.

```ts
// TypeScript errors on overtly wrong ways of populating the `void` type...
function returnsVoid(): void {
    return 'not void!'; // TS Error:  Type 'string' is not assignable to type 'void'.
}

// ...but allows more subtle ones
const returnsVoid: () => void = () => 1234;

// Likewise, TypeScript errors on overtly wrong usages of `void` as a runtime value...
declare const v: void;
if (v) { // TS Error: An expression of type 'void' cannot be tested for truthiness.
    // ... do something
}

// ...but silently treats `void` as `undefined` in more subtle scenarios
declare const voidOrString: void | string;
if (voidOrString) {
     // voidOrString is narrowed to string in this branch, so this is allowed.
     console.log(voidOrString.toUpperCase());
}
```

Putting these behaviors together, we can construct a scenario where TS will throw at runtime, despite not reporting a type error:

````

JoshuaKGoldberg commented:
👋 @phaux is this ready for re-review?

phaux commented:
Not yet, but I am thinking every day about how to simplify messages in this PR

JoshuaKGoldberg commented:
I'll switch this to draft then 👍.

JoshuaKGoldberg commented:
Checking in again @phaux - is this ready for review? Anything we can help with?

phaux commented:
I addressed most of the issues. 

I left some unrelated refactors in another rule because they were needed for this rule's auto fixes, which are now removed and will be sent as separate PR. If that's important I can try to remove the unrelated changes too for now. I have to do it manually because they were part of the big initial commit and I was lazy.

The messages still use dynamic IDs, but I reduced their number by a lot. If that's still too many messages, let me know and I will just replace all of them with a single generic message like "value returned in void context" and move them to a separate PR so it can be discussed later too.

JoshuaKGoldberg commented on packages/eslint-plugin/src/util/discardReturnValueFix.ts:
```diff
@@ -0,0 +1,91 @@
+  if (newReturnText[0] === '{') {
+    // The value would be interpreted as a block statement,
+    // so we need to wrap it in parentheses.
+    newReturnText = `(${newReturnText})`;
```
[Testing] This isn't covered by any tests - either it's a valid case that should be tested, or it's not and we can remove it.

JoshuaKGoldberg commented on packages/utils/src/eslint-utils/assert.ts:
[Refactor] -1 from me in general on this extra set of assertions. There's a reason we haven't leaned into this pattern more: values should be their expected type before being used. 

E.g. instead of:

```ts
memberTsNode: ts.MethodDeclaration | ts.PropertyDeclaration,
// ...
ESLintUtils.assert(ts.isClassLike(memberTsNode.parent), '...');
```

...we'd want either:

* If `ts.MethodDeclaration | ts.PropertyDeclaration`'s `.parent` _is_ always a `ts.ClassLikeDeclaration`, the AST's types should reflect that
* If `ts.MethodDeclaration | ts.PropertyDeclaration`'s `.parent` _isn't_ always a `ts.ClassLikeDeclaration`, then `memberTsNode`'s type should have something like `{ parent: ts.ClassLikeDeclaration }` added

In other words: I think these assertions are surface-level fixes for type issues that should be fixed at a deeper level.

---

I'll also note that we generally try to work with the TSESTree AST rather than TypeScript's, when possible. Will mention this on `getBaseTypesOfClassMember` itself.

JoshuaKGoldberg commented on packages/eslint-plugin/src/util/getBaseTypesOfClassMember.ts:
```diff
@@ -0,0 +1,42 @@
+ */
+export function* getBaseTypesOfClassMember(
+  checker: ts.TypeChecker,
+  memberTsNode: ts.MethodDeclaration | ts.PropertyDeclaration,
```
Continuing from the comment on assertions: I'll also note that we generally try to work with the TSESTree AST rather than TypeScript's, when possible. We have more flexibility and power to make the AST more specific because we're the ones who define it (and we define it as a discriminated union).

In this case it'd be more typical for a typed lint rule to take in the TSESTree node. You can also pass in the `services: ParserServicesWithTypeInformation` to use APIs like `services.getSymbolAtLocation`.

phaux replied:
Done

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,636 @@
+      asyncFuncInAttr:
+        'Async event handler `{{attrName}}` passed as a prop to `{{elemName}}`, which expects a void event handler.',
+      asyncFuncInExtMember:
+        'Async function provided as `{{memberName}}` method of `{{className}}`, whose base class `{{baseName}}` declares it as a void method.',
```
[Typo] Grammar fix:

```suggestion
        'Async function provided as `{{memberName}}` method of `{{className}}`, whose base class `{{baseName}}` declares as a void method.',
```

JoshuaKGoldberg commented on packages/eslint-plugin/tests/rules/strict-void-return.test.ts:
[Testing] A few more edge cases to test

* `extends Bar implements Foo, Baz`
* `extends class { ... }` and  `extends class WithName { ... }`
* `implements class { ... }` and `implements class WithName { ... }`

phaux replied:
`class implements class {} {}` is already invalid TS

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,636 @@
+    },
+    fixable: 'code',
+    hasSuggestions: false,
+    messages: {
```
From https://github.com/typescript-eslint/typescript-eslint/pull/9707#issuecomment-2575990827:

> The messages still use dynamic IDs, but I reduced their number by a lot. If that's still too many messages, let me know and I will just replace all of them with a single generic message like "value returned in void context" and move them to a separate PR so it can be discussed later too.

For me, dynamic IDs are a hard blocker. All IDs should be statically constructed for readability - both for us humans and the tools around them.

I'm not opposed to this PR having more specific reports for different cases. I think a few of the delineations are helpful:

* Declaring a class/object method/property _vs._ providing an argument to a function
* Explicitly stating which base class or interface made a method/property report

Looking at the code, it's ~7 archetypes of messages repeated across each of 3 groups. How about reducing to just 7 messages overall?

phaux replied:
This wont work because the message id and parts are decided at the beginning and they would still need to be passed to `reportIfNonVoidFunction` via argument and would be dynamic anyways.

So there can't be "where" distinction (variable / argument / class member), only the "what" distinction (return / async func / non-void func), because it's computed at the end. I agree the first is the more useful one.

I will remove all messages and leave a single generic one.

phaux replied:
**Unless** we remove the "what" distinction and the whole `reportIfNonVoidFunction` so that every node check calls `context.report` directly with a static messageId.

But that distinction will be necessary for the autofixes anyways so it's not possible neither.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-confusing-void-expression.ts:
From https://github.com/typescript-eslint/typescript-eslint/pull/9707#issuecomment-2575990827:

> I left some unrelated refactors in another rule because they were needed for this rule's auto fixes, which are now removed and will be sent as separate PR. If that's important I can try to remove the unrelated changes too for now. I have to do it manually because they were part of the big initial commit and I was lazy.

Yes, please revert any changes that aren't necessary for the specific added rule in this PR.

FWIW this is part of why we don't require rebasing and actually request not to force-push in https://typescript-eslint.io/contributing/pull-requests 😄. We don't want to add "keeping the Git history pristine" to the tasks for you the PR author or us as maintainers.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,636 @@
+
+        if (!options.allowReturnPromiseIfTryCatch) {
+          // Async functions aren't allowed.
+          // TODO: Suggest wrapping its body in an async IIFE.
```
[Docs] Could you remove these TODOs actually? We haven't decided what the right next step would be. I'm personally loathe to suggest async IIFEs; IMO they're 99% of the time an anti-pattern that results in dangerous async patterns.

phaux replied:
I agree that AIIFEs suck, but that's the only way (that I know) to make this rule (or no-misused-promises) happy. That's why I added that suggestion and ignoreAsyncInTryCatch.

phaux replied:
Removed

JoshuaKGoldberg commented on packages/eslint-plugin/tests/rules/strict-void-return.test.ts:
```diff
@@ -0,0 +1,2767 @@
+      ],
+      options: [{ allowReturnUndefined: false }],
+    },
+    // TODO: Check every union type separately
```
[Question] Is this still pending? What's up?

phaux replied:
This one was non-trivial to implement and I think it's not worth it.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,636 @@
+  defaultOptions: [
+    {
+      allowReturnNull: false,
+      allowReturnPromiseIfTryCatch: true,
```
[Feature] In this PR's OP, `allowReturnPromiseIfTryCatch` is only mentioned in the collapsed note around fixers & suggestions. Seems to me it's no longer necessary, right? 

ACK that it's not much added code but I don't think we have consensus that it's a strategy we'd want to add to the rule. Could you please split it out? We can always talk about it more in a followup issue - either its own and/or the one around fixers & suggestions.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,636 @@
+        'Disallow passing a value-returning function in a position accepting a void function',
+      requiresTypeChecking: true,
+    },
+    fixable: 'code',
+    hasSuggestions: false,
```
```suggestion
```

phaux replied:
fixed

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,636 @@
+    {
+      allowReturnNull: false,
+      allowReturnPromiseIfTryCatch: true,
+      allowReturnUndefined: true,
```
[Question] Separate from the promise/try-catch options, I don't know that we really need the remaining `allowReturn*` options. They're added complexity for the first version of a rule we haven't tried out in the real world yet. Do you see a strong need for them? (if not, I'm -1 on including them)

phaux replied:
removed

JoshuaKGoldberg reviewed (changes requested):
Progress! 🚀 

Requesting some more simplification changes that'll help us review please.

JoshuaKGoldberg commented:
👋 Checking in @phaux, do you think you'll have the time to drive this to completion over the next month or two? I really like the idea of the rule but it's been sitting for a while - it'd be nice to either really push it forward or open it up to get worked on.

phaux commented:
@JoshuaKGoldberg Sorry for the wait. I recreated this branch from scratch and the next version will be as minimal as possible so hopefully it finally passes review :) I will send it in few days probably.

phaux commented:
Ready

phaux commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,452 @@
+      JSXAttribute: (node): void => {
+        if (
+          node.value?.type === AST_NODE_TYPES.JSXExpressionContainer &&
+          node.value.expression.type !== AST_NODE_TYPES.JSXEmptyExpression
```
Added valid tests for JSXEmptyExpression (`prop={}`)  and JSXElement (`prop=<elem/>`) but I dunno how to get JSXSpreadChild. 
I tried `prop={...value}` but it doesn't work.

phaux commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,452 @@
+        .flatMap(signature =>
+          tsutils.unionTypeParts(signature.getReturnType()),
+        );
+      return (
```
This is just a coincidence. I simplified this condition so that it doesn't work for the other function anymore and now they are different. I think it makes sense to keep them separate.

phaux commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,452 @@
+      return (
+        // At least one return type is void
+        returnTypes.some(type =>
+          tsutils.isTypeFlagSet(type, ts.TypeFlags.Void),
```
dunno

phaux commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,452 @@
+    function checkObjectPropertyNode(
+      propNode: TSESTree.MethodDefinition | TSESTree.Property,
+    ): void {
+      if (
+        propNode.value.type === AST_NODE_TYPES.AssignmentPattern ||
+        propNode.value.type === AST_NODE_TYPES.TSEmptyBodyFunctionExpression
+      ) {
+        return;
+      }
```
I think in this case the types are wrong. 

AFAIK it's not possible to have TSEmptyBodyFunctionExpression as an object's property value. `{ f() }` is an syntax error. This is only valid in an abstract class body.

AssignmentPattern is also not valid as value of Property of ObjectLiteral, but it is valid for Property of ObjectPattern. Types could be improved by splitting Property into 2 types.

I will add type assertions for now.

phaux commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,452 @@
+    function checkClassPropertyNode(
+      propNode: TSESTree.PropertyDefinition,
+    ): void {
+      if (propNode.value == null) {
+        return;
+      }
```
There is already a test for this and if I remove it then the types are wrong. TS functions say they don't accept null but if you pass null they just do nothing and return more null so nothing crashes.

phaux commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,452 @@
+     * and the implemented interfaces (when the class implements an interface).
+     */
+    function checkClassMethodNode(methodNode: TSESTree.MethodDefinition): void {
+      if (
+        methodNode.value.type === AST_NODE_TYPES.TSEmptyBodyFunctionExpression
+      ) {
+        return;
+      }
+      if (methodNode.kind !== 'method') {
+        return;
+      }
```
Added test for empty body func.

Removing `kind == 'method'` actually makes the rule catch more cases.

phaux commented on packages/eslint-plugin/docs/rules/strict-void-return.mdx:
```diff
@@ -0,0 +1,416 @@
+</TabItem>
+</Tabs>
+
+### Void context examples
```
Removed and changed other headings slightly

phaux replied:
I can make a blog post WIP PR directly later if you don't mind. It would contain mostly the part I removed for now.

phaux commented on packages/eslint-plugin/docs/rules/strict-void-return.mdx:
```diff
@@ -0,0 +1,416 @@
+</TabItem>
+</Tabs>
+
+### Void context from base class
```
I already changed it but I just remembered my reasoning: You can only extend one class in a class declaration (unlike many interfaces) that's why this one was singular :laughing:

phaux commented:
I need help with some of the review comments

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,428 @@
+              'Whether to allow functions returning `any` to be used in place expecting a `void` function.',
+          },
+        },
+      },
```
[Feature] This is what [`@typescript-eslint/no-unsafe-return`](https://typescript-eslint.io/rules/no-unsafe-return) already does, right? I don't think we need this rule to also check for `any` safety - it's not the same area as strict `void` semantics. I think this rule should effectively hardcode `allowReturnAny` to enabled and not let users change that.

```suggestion
```

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,428 @@
+      },
+      PropertyDefinition: checkClassPropertyNode,
+      ReturnStatement: (node): void => {
+        if (node.argument != null) {
```
[Testing] Removing this `if` and just calling `checkExpressionNode(node.argument);` doesn't fail any unit tests. Unnecessary code, or missing tests?

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/strict-void-return.ts:
```diff
@@ -0,0 +1,428 @@
+      for (const [argIdx, argNode] of callNode.arguments.entries()) {
+        if (argNode.type === AST_NODE_TYPES.SpreadElement) {
+          continue;
+        }
```
[Testing] Removing this `if` altogether doesn't fail any unit tests. Unnecessary code, or missing tests?

JoshuaKGoldberg replied:
Heh, I tried for 10 minutes and couldn't get this to work either. Tricky!

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/strict-void-return.mdx:
```diff
@@ -0,0 +1,321 @@
+## When Not To Use It
+
+Some projects are architected so that values returned from synchronous void functions are generally safe.
+If you only want to check for misused voids with asynchronous functions then you can use [`no-misused-promises`](./no-misused-promises.mdx) instead.
```
```suggestion
If you only want to check for misused `void`s with asynchronous functions then you can use [`no-misused-promises`](./no-misused-promises.mdx) instead.
```

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/strict-void-return.mdx:
```diff
@@ -0,0 +1,321 @@
+
+## When Not To Use It
+
+Some projects are architected so that values returned from synchronous void functions are generally safe.
```
```suggestion
Some projects are architected so that values returned from synchronous `void` functions are generally safe.
```

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/strict-void-return.mdx:
```diff
@@ -0,0 +1,321 @@
+Some projects are architected so that values returned from synchronous void functions are generally safe.
+If you only want to check for misused voids with asynchronous functions then you can use [`no-misused-promises`](./no-misused-promises.mdx) instead.
+
+In browser context, an unhandled promise will be reported as an error in the console.
```
```suggestion
In browser contexts, an unhandled Promise will be reported as an error in the console.
```

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/strict-void-return.mdx:
```diff
@@ -0,0 +1,321 @@
+but if you are just prototyping or don't care about that, the default behavior might be acceptable.
+In such case, instead of handling the promises and `console.error`ing them anyways, you can just disable this rule.
+
+Similarly, the default behavior of crashing the process on unhandled promise rejection
```
```suggestion
Similarly, the default server runtime behaviors of crashing the process on unhandled Promise rejections
```

JoshuaKGoldberg reviewed (dismissed):
Ok! I must apologize for another long gap in reviewing. This rule intimidated me and I procrastinated on it far too long. Again, sorry.

I think it's roughly ready for merge. It's big and intricate enough that I'm positive we'll get a bunch of issues filed once people start using it. But I think a gradual rollout to let folks try it will make sense:

1. Merge roughly as-is, not in any preset configs
2. We can put a call out on social media for folks to try the rule
3. Eventually we can add it to `strictTypeChecked` - which'll get a lot of additional testing (recent prior art: `no-useless-default-assignment`)

I left some comments on things I think should block merge, but I feel comfortable applying them myself to save you the time (especially the unit tests splitting-up request). I'll just wait for another review from @typescript-eslint/triage-team.

Thanks again! ❤️
